stl/src/vector_algorithms.cpp

```diff
@@ -2930,10 +2930,378 @@ namespace {
+            if constexpr (sizeof(_Ty) == 1) {
+                return true;
+            } else {
+                __m128i _Mask = _mm_undefined_si128();
+                if constexpr (sizeof(_Ty) == 2) {
+                    _Mask = _mm_set1_epi16(static_cast<short>(0xFF00));
+                } else if constexpr (sizeof(_Ty) == 4) {
+                    _Mask = _mm_set1_epi32(static_cast<int>(0xFFFF'FF00));
+                } else if constexpr (sizeof(_Ty) == 8) {
+                    _Mask = _mm_set1_epi64x(static_cast<long long>(0xFFFF'FFFF'FFFF'FF00));
+                } else {
+                    static_assert(false, "Unexpected size");
+                }
+
+                const size_t _Byte_size = _Needle_length * sizeof(_Ty);
+
+                const void* _Stop = _Needle_ptr;
+                _Advance_bytes(_Stop, _Byte_size & ~size_t{0x1F});
+                while (_Needle_ptr != _Stop) {
+                    const __m128i _Data = _mm_loadu_si128(reinterpret_cast<const __m128i*>(_Needle_ptr));
+                    if (!_mm_testz_si128(_Mask, _Data)) {
+                        return false;
+                    }
+
+                    _Needle_ptr += 32 / sizeof(_Ty);
+                }
+
+                _Advance_bytes(_Stop, _Byte_size & 0x1E);
+                while (_Needle_ptr != _Stop) {
+                    if ((*_Needle_ptr & ~_Ty{0xFF}) != 0) {
+                        return false;
+                    }
+
+                    ++_Needle_ptr;
+                }
+
+                return true;
+            }
+        }
+
+        __m256i _Step(const __m256i _Bitmap, const __m256i _Data) noexcept {
+            const __m256i _Data_high    = _mm256_srli_epi32(_Data, 5);
+            const __m256i _Bitmap_parts = _mm256_permutevar8x32_epi32(_Bitmap, _Data_high);
+            const __m256i _Data_low_inv = _mm256_andnot_si256(_Data, _mm256_set1_epi32(0x1F));
+            const __m256i _Mask         = _mm256_sllv_epi32(_Bitmap_parts, _Data_low_inv);
+            return _Mask;
+        }
+
+        template <class _Ty>
+        __m256i _Load_avx_256_8(const _Ty* const _Src) noexcept {
+            if constexpr (sizeof(_Ty) == 1) {
+                return _mm256_cvtepu8_epi32(_mm_loadu_si64(_Src));
+            } else if constexpr (sizeof(_Ty) == 2) {
+                return _mm256_cvtepu16_epi32(_mm_loadu_si128(reinterpret_cast<const __m128i*>(_Src)));
+            } else if constexpr (sizeof(_Ty) == 4) {
+                return _mm256_loadu_si256(reinterpret_cast<const __m256i*>(_Src));
+            } else if constexpr (sizeof(_Ty) == 8) {
+                const __m256i _Low  = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(_Src));
+                const __m256i _High = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(_Src) + 1);
+                const __m256i _Pack = _mm256_packs_epi32(_Low, _High);
+                return _mm256_permutex_epi64(_Pack, _MM_SHUFFLE(3, 1, 2, 0));

```

Comment:
Question: According to the Intel Intrinsics Guide, [`_mm256_permutex_epi64`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_permutex_epi64&ig_expand=4957) is:

```
__m256i _mm256_permutex_epi64 (__m256i a, const int imm8)
#include <immintrin.h>
Instruction: vpermq ymm, ymm, imm8
CPUID Flags: AVX512F + AVX512VL
```

Does this require AVX512 support?

Reply:
It works for me, and I don't have AVX512.

This intrinsic exists as maskless form of AVX512 masked `vpermq`, but the compiler emits AVX2 form for that.

However, let's use `_mm256_permute4x64_epi64` form instead, that is documented to be AVX2.